<?xml version="1.0" encoding="utf-8"?>
<launch>
    <arg name="robot_prefix" default="/robot5" />
    <arg name="input_points_topic" default="/odometry_projected_cloud" />
    <arg name="output_pointcloud_topic_name" default="/odometry_projected_cloud" />
    <arg name="output_laser_link" default="$(arg robot_prefix)/calibrated_velodyne_link" />
    <arg name="sensor_offset_topic" default="/sensor_default" />
    <arg name="odom_link" default="odom" />
    <arg name="base_link" default="/base_link" />
    <arg name="world_link" default="/world" />
    <arg name="sensor_pose_x" default="0" />
    <arg name="sensor_pose_y" default="0" />
    <arg name="sensor_pose_z" default="0" />
    <arg name="sensor_pose_r" default="0" />
    <arg name="sensor_pose_p" default="0" />
    <arg name="sensor_pose_t" default="0" />
    <arg name="laser_2d" default="false" />
    <arg name="output" default="screen" />


    <!-- start mapping node -->
    <node pkg="graph_map" type="graph_map_fuser_node" name="graph_node" output="$(arg output)">

        <!-- Choose weather to initiate pose to pose_init_<x,y,t> or the data of /<gt_topic> -->
        <param name="initPoseFromGT" value="false" />
        <param name="pose_init_x" value="0" />
        <param name="pose_init_y" value="0" />
        <param name="pose_init_t" value="0" />


        <!-- Select mapping technique to one of the methods listed in graph_map/include/graph_factory.h -->
        <param name="map_type" value="ndt_map" />

        <!-- use local maps instead of a single map, this can improve map descriptiveness -->
        <param name="use_submap" value="false" />
        <!-- choose to disable essential parts of the software -->
        <param name="enable_registration" value="false" />

        <!-- Select registration technique to one of the methods listed in graph_map/include/graph_factory.h -->
        <param name="registration_type" value="ndt_d2d_reg" />
        <param name="enable_mapping" value="false" />

        <param name="tf_pose_frame" value="" />

        <!-- if laser_2d is true, a 2d scanner will be used for mapping-->
        <param name="laser_2d" value="$(arg laser_2d)" />

        <param name="points_topic" value="$(arg input_points_topic)" />
        <param name="output_pointcloud_topic_name" value="$(arg output_pointcloud_topic_name)" />

        <param name="useOdometry" value="false" />
        <param name="odometry_topic" value="$(arg robot_prefix)/control/odom" />
        <param name="use_tf_listener" value="true" />
        <param name="use_pointcloud" value="true" />

        <param name="gt_topic" value="$(arg robot_prefix)/control/state" />

        <param name="publish_laser_link" value="true" />
        <param name="laser_frame_id" value="$(arg output_laser_link)" />

        <!--  odometry frame id-->
        <param name="world_frame" value="$(arg world_link)" />
        <param name="use_REP_convention" value="true" />
        <param name="odom_tf" value="$(arg robot_prefix)/$(arg odom_link)" />
        <param name="base_tf" value="$(arg robot_prefix)/$(arg base_link)" />

        <!-- size of map -->
        <param name="size_x_meters" value="70" />
        <param name="size_y_meters" value="70" />
        <param name="size_z_meters" value="10" />
        <param name="sensor_offset_topic" value="$(arg sensor_offset_topic)" />

        <param name="laser_variance_z" value="0.02" />

        <!--range of sensor, min range can prevent self-mapping of the robot or operator of the lidar, max range can remove uncertain measurments -->
        <param name="max_range" value="130." />
        <param name="min_range" value="2.5" />

        <!-- Specific sensor offset parameters with respect to the odometry frame -->

        <param name="sensor_pose_x" value="$(arg sensor_pose_x)" />
        <param name="sensor_pose_y" value="$(arg sensor_pose_y)" />
        <param name="sensor_pose_z" value="$(arg sensor_pose_z)" />
        <param name="sensor_pose_r" value="$(arg sensor_pose_r)" />
        <param name="sensor_pose_p" value="$(arg sensor_pose_p)" />
        <param name="sensor_pose_t" value="$(arg sensor_pose_t)" />

        <param name="visualize" value="false" />
        <param name="use_keyframe" value="false" />

    </node>


</launch>